### PR TITLE
Implement a lazy-messages mechanism

### PR DESCRIPTION
**Pull request by DAB0mB**
- Implement a lazy-messages mechanism and apply it to messages view
- Replace native document usage with ElementRef service
- Add comments to messages component
- Add day groups to messages page so we can know which day the messages were sent

### CODE REVIEW

File: src/pages/messages/messages.ts [hunk 1]
```diff
@@ -57,40 +59,135 @@ export class MessagesPage implements OnInit, OnDestroy {
...
+  subscribeMessages(): Subscription {
+    // Prohibit parallel subscriptions
+    if (this.loadingMessages) return;
+    // A flag which indicates if there's a subscription in process
+    this.loadingMessages = true;
+    // A custom offset to be used to re-adjust the scrolling position once
+    // new dataset is fetched
+    this.scrollOffset = this.scroller.scrollHeight;
+
+    return MeteorObservable.subscribe('messages',
```
Comment (dotansimha) [inline src/pages/messages/messages.ts]: You return this `Observable` but no one uses it. 
I think you can use it and I think that then you can remove `loadingMessages` boolean...

File: src/pages/messages/messages.ts [hunk 2]
```diff
@@ -57,40 +59,135 @@ export class MessagesPage implements OnInit, OnDestroy {
...
+    this.subscribeMessages();
+
+    // Get total messages count in database so we can have an indication of when to
+    // stop the auto-subscriber
+    MeteorObservable.call('countMessages').subscribe((messagesCount: number) => {
+      // Note that the 'scroller' element is being created dynamically by an Ionic
+      // component and therefore the event listener can't be registered directly
+      // from the view
+      let scrollListener = this.onScroll.bind(this);
+      this.scroller.addEventListener('scroll', scrollListener);
```
Comment (dotansimha) [inline src/pages/messages/messages.ts]: Maybe you can wrap the scroll events with `Observable` and then treat the scroll changes as a stream.

File: src/pages/messages/messages.ts [hunk 3]
```diff
@@ -57,40 +59,135 @@ export class MessagesPage implements OnInit, OnDestroy {
...
+  ): Subscription {
+    return MeteorObservable.autorun().subscribe(() => {
+      if (messagesCount != Messages.collection.find().count()) return;
+      this.scroller.removeEventListener('scroll', scrollListener);
+    });
+  }
+
+  // Detects changes in the messages dataset and re-renders the view
+  autorunMessages(): Subscription {
+    return MeteorObservable.autorun().subscribe(() => {
```
Comment (dotansimha) [inline src/pages/messages/messages.ts]: Why use `autorun`? you can use the collection to detect changes...

File: src/pages/messages/messages.ts (hunk 1 above)
Comment (DAB0mB) [inline src/pages/messages/messages.ts]: The `loadingMessages` boolean is used like some sort of a mutex where it prevents from parallel subscription requests to be done all at the same time which can potentially overflow the Meteor server with unnecessary requests.

File: src/pages/messages/messages.ts (hunk 3 above)
Comment (DAB0mB) [inline src/pages/messages/messages.ts]: The `autorun` is used to re-render the view in-case there are incoming messages from the receiver. What do you mean by "use the collection"? How exactly do you want me to use it?

File: src/pages/messages/messages.ts (hunk 2 above)
Comment (DAB0mB) [inline src/pages/messages/messages.ts]: This is a single event registration while the `Observable` is used to handle multiple async tasks, hence I decided not to use it

Comment (dotansimha) [inline src/pages/messages/messages.ts]: You can replace the `scrollListener` method with `Observable` logic, and then use is along with `combineLatest`...

File: src/pages/messages/messages.ts (hunk 1 above)
Comment (dotansimha) [inline src/pages/messages/messages.ts]: If you debounce the scroll events using `Observable`, use the `Subscription` return value and save it on your instance (`this`) - then check if this variable is `undefined`... then you won't need the boolean

File: src/pages/messages/messages.ts (hunk 3 above)
Comment (dotansimha) [inline src/pages/messages/messages.ts]: I see. you can also use the `Observable` you get from the cursor (`...find()`), and subscribe to it in order to know when you collection changes.

**Comment by DAB0mB** [comment]
@dotansimha I fixed everything you told me except the `loadingMessage` flag; there are many ways to determine whether the messages subscription is being loaded or not, but I think this once is the most explicit and the most clear of all :-) I'd also like to note that there is a problem with the `ObservableCursor.collectionCount()` exported by the `meteor-rxjs` library.

**Comment by DAB0mB** [comment]
Also I don't think you noticed but the `loadingMessages` flag is being used in the `scrollDown` method, so it's super important not only for the subscription, but also for scrolling.

**Pull request merged**